### PR TITLE
feat(ec2): CMK support for EBS Volume encryption

### LINKED ISSUE #6459: Encrypted EBS volumes for EC2 instances
Support encrypted EBS volumes for EC2 instances.

### Use Case

We are using the CDK to generate EC2 fleets with EBS volumes that will contain customer sensitive information. We want customers to be able to use AMIs that we make available; the way to have a public AMI launch an encrypted instance is through this new EBS functionality.

### Proposed Solution

I think that the option is via AWS::EC2::Instance → BlockDeviceMappings→ebs→ Encrypted/KmsKeyId, or 
AWS::EC2::LaunchTempate→BlockDeviceMappings..., or 
AWS::EC2::SpotFleet→ BlockDeviceMappings...

There does not appear to be a way to specify information that would feed into the BlockDeviceMapping in the L2 ec2.Instance construct

### Other

Details: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AMIEncryption.html

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by alles-klar**
Added support for kmsKeyId in blockDevices to use KMS customer master key (CMK) for EBS encryption.

Fixes #6459

----

*This PR is a copy of https://github.com/aws/aws-cdk/pull/8172 because the author does not respond anymore. But all credits for the initial work belong to @shreyasdamle*

----
On behalf of DB Systel GmbH

### CODE REVIEW

**Comment by alles-klar** [comment]
Feedback from @iliapolo is implemented.

File: packages/@aws-cdk/aws-ec2/lib/volume.ts [hunk 1]
```diff
@@ -129,6 +157,17 @@ export class BlockDeviceVolume {
    * @param options additional device options
    */
   public static ebs(volumeSize: number, options: EbsDeviceOptions = {}): BlockDeviceVolume {
+    // If KmsKeyId is specified, the encrypted state must be true.
+    if (options.kmsKeyId) {
+      if (options.encrypted === false) {
+        throw new Error('`encrypted` must be not false when providing `kmsKeyId`.');
+      } else {
+        // Encryption is implied if KMS key is specified.
+        (options.encrypted as unknown as boolean) = true;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/volume.ts:166]: Please don't do this. Instead, reassign a new object that has the values you want to set:

```ts
options = {
   ...options,
  encrypted: true,
  kmsKeyId: options.kmsKey.keyArn,
};
```

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-ec2/lib/volume.ts [hunk 2]
```diff
@@ -90,6 +90,34 @@ export interface EbsDeviceOptions extends EbsDeviceOptionsBase {
...
+   *         "StringEquals": {
+   *           "kms:ViaService": "ec2.<Region>.amazonaws.com", (eg: ec2.us-east-1.amazonaws.com)
+   *           "kms:CallerAccount": "0000000000" (your account ID)
+   *         }
+   *       }
+   *     }
+   *
+   * @default None
+   */
+  readonly kmsKeyId?: IKey;
```
[the remark this answers is not in the document]
Comment (alles-klar) [reply packages/@aws-cdk/aws-ec2/lib/volume.ts]: I can change kmsKey it but then there is a braking change/new parameter name in the volume class. If it doesn't matter, I will change it when everything else is working.

**Comment by alles-klar** [comment]
Sorry for the delay. I tried to fix the code.

Unfortunately two tests fail now. I spend some hours of debugging an it is related to the kmsKey but I have no idea what is wrong. Maybe a permission problem? I am getting the following error:

```
blockDeviceMappings › can set customer managed key

    Resolution error: Resolution error: Unable to resolve object tree with circular reference. Path: /Resources/${Token[Default.Instance.Resource.LogicalID.863]}/Properties/blockDeviceMappings/0/ebs/kmsKey/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host..

      59 |    * Whether or not to allow intrinsics in keys of an object
      60 |    *
    > 61 |    * Because keys of an object must be strings, a (resolved) intrinsic, which
         |           ^
      62 |    * is an object, cannot be stored in that position. By default, we reject these
      63 |    * intrinsics if we encounter them.
      64 |    *

      at resolve (../core/lib/private/resolve.ts:61:11)
      at Object.resolve (../core/lib/private/resolve.ts:53:32)
      at resolve (../core/lib/private/resolve.ts:165:39)
      at Object.resolve (../core/lib/private/resolve.ts:53:32)
      at resolve (../core/lib/private/resolve.ts:165:39)
      at Object.resolve (../core/lib/private/resolve.ts:53:32)
      at resolve (../core/lib/private/resolve.ts:165:39)
      at Object.resolve (../core/lib/private/resolve.ts:53:32)
      at resolve (../core/lib/private/resolve.ts:165:39)
      at Object.resolve (../core/lib/private/resolve.ts:53:32)
```

@rix0rrr Do you have an idea where the error comes from?

File: packages/@aws-cdk/aws-ec2/lib/volume.ts (hunk 1 above)
Comment (alles-klar) [reply packages/@aws-cdk/aws-ec2/lib/volume.ts]: Fixed but I the tests fail now... -> https://github.com/aws/aws-cdk/pull/12842#issuecomment-801987291

**Review by rix0rrr — APPROVED** [review]

File: packages/@aws-cdk/aws-ec2/lib/private/ebs-util.ts [hunk 3]
```diff
@@ -34,9 +35,19 @@ function synthesizeBlockDeviceMappings<RT, NDT>(construct: Construct, blockDevic
       } else if (volumeType !== EbsDeviceVolumeType.IO1) {
         Annotations.of(construct).addWarning('iops will be ignored without volumeType: EbsDeviceVolumeType.IO1');
       }
+
+      cfnEbs = {
+        ...ebs,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/private/ebs-util.ts:40]: Ah hah.

The problem is that this object still contains the `kmsKey` object, which is why the tests are failing with this error:

```
@aws-cdk/aws-ec2:     Resolution error: Resolution error: Unable to resolve object tree with circular reference. Path: /Resources/${Token[Default.Instance.Resource.LogicalID.983]}/Properties/blockDeviceMappings/0/ebs/kmsKey/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host..
```

It seems like the fields were previously INCIDENTALLY identical to the `CfnEbs` object, but no longer. You need to copy the fields out individually (and transform them as necessary).

**Review by rix0rrr — CHANGES_REQUESTED** [review]

**Comment by alles-klar** [comment]
@rix0rrr tests are green now, but the pipeline is failing because of breaking API changes. Are you ok with these changes?

**Comment by rix0rrr** [comment]
We're never okay with breaking changes. Let's see what they are...

**Review by rix0rrr — CHANGES_REQUESTED** [review]
These need to be addressed:

```
err  - IFACE @aws-cdk/aws-ec2.EbsDeviceOptions: has been removed [removed:@aws-cdk/aws-ec2.EbsDeviceOptions]
err  - METHOD @aws-cdk/aws-ec2.BlockDeviceVolume.ebs: argument options, takes Optional<@aws-cdk/aws-ec2.EbsDeviceOptionsBase> (formerly Optional<@aws-cdk/aws-ec2.EbsDeviceOptions>): could not find type @aws-cdk/aws-ec2.EbsDeviceOptions [incompatible-argument:@aws-cdk/aws-ec2.BlockDeviceVolume.ebs]
```

Can't get rid of `EbsDeviceOptions`.

**Pull request closed without merging**